pg_search/src/query/proximity/weight.rs

```diff
@@ -0,0 +1,212 @@
+                self.fieldnorm_reader(segment_reader)?,
+                self.weight_opt.clone(),
+            );
+
+            let mut doc_ids = Vec::new();
+            let mut positions = Vec::new();
+            let mut offsets = Vec::new();
+            while scorer.doc() != TERMINATED {
+                offsets.push(positions.len() as u32);
+                doc_ids.push(scorer.doc());
+
+                for (l, r) in scorer.prox_iter() {
+                    if nested {
+                        match which_terms {
+                            // *** NOTICE! ***
+                            //
+                            // these are *purposely* reversed.  If we're nested and the user is asking
+                            // for postings on the left-side of a proximity query what *we* need to
+                            // return are the postings from the right side
+                            WhichTerms::Left => positions.push(r),
+
+                            // same goes for the right-side, just in reverse
+                            WhichTerms::Right => positions.push(l),
+
+                            // gotta collect 'em all
+                            WhichTerms::All => {
+                                positions.push(l);
+                                positions.push(r);
+                            }
+                        }
+                    } else {
+                        positions.push(l);
+                        positions.push(r);
+                    }
+                }
+                scorer.advance();
+            }
+            offsets.push(positions.len() as u32);
+
+            let loaded_postings = LoadedPostings {
+                doc_ids: doc_ids.into_boxed_slice(),
+                position_offsets: offsets.into_boxed_slice(),
+                positions: positions.into_boxed_slice(),
+                cursor: 0,
+            };
+
+            Ok(vec![Box::new(loaded_postings)])
+        } else {
+            let mut postings: Vec<Box<dyn Postings>> = Vec::new();
+            let mut num_regex_terms = 0;
+            let inverted_index = segment_reader.inverted_index(self.query.field())?;
+            for term in clause.terms(self.query.field(), Some(segment_reader), which_terms)? {
+                match term.as_ref() {
+                    ProximityTermStyle::Term(term) => {
+                        let term = Term::from_field_text(self.query.field(), term);
+                        if let Some(segment_postings) = inverted_index
+                            .read_postings(&term, IndexRecordOption::WithFreqsAndPositions)?
+                        {
+                            postings.push(Box::new(segment_postings));
+                        }
+                    }
+                    ProximityTermStyle::Rexgex(re, max_expansions) => {
+                        let regex =
+                            tantivy_fst::Regex::new(re.as_str()).unwrap_or_else(|e| panic!("{e}"));
```

Comment:
Likely premature, but: would it be possible to combine all of the Regex terms into a single automaton to find postings in one pass?